### PR TITLE
stat/distmv: add Renyi divergence for Gaussians

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

File: stat/distmv/statdist.go [hunk 1]
```diff
@@ -207,6 +207,76 @@ func (KullbackLeibler) DistUniform(l, r *Uniform) float64 {
 	return logPx - logQx
 }
 
+// Renyi is a type for computing the Rényi divergence of order α from l to r.
+// The dimensions of the input distributions must match or the function will panic.
+//
+// The Rényi divergence with α > 0, α ≠ 1 is defined as
+//  D_α(l || r) = 1/(α-1) log(\int_x l(x)^α r(x)^(1-α)dx)
```
Comment (vladimir-ch) [inline stat/distmv/statdist.go]: What is the interval of integration? Using the integration variable there is not good, so writing just `\int` is better.  The same for Bhattacharyya.

File: stat/distmv/statdist.go [hunk 2]
```diff
@@ -15,7 +15,7 @@ import (
 // Bhattacharyya is a type for computing the Bhattacharyya distance between
 // probability distributions.
 //
-// The Battachara distance is defined as
+// The Battacharyya distance is defined as
```
Comment (vladimir-ch) [inline stat/distmv/statdist.go]: Bhattacharyya

File: stat/distmv/statdist.go [hunk 3]
```diff
@@ -207,6 +207,76 @@ func (KullbackLeibler) DistUniform(l, r *Uniform) float64 {
 	return logPx - logQx
 }
 
+// Renyi is a type for computing the Rényi divergence of order α from l to r.
+// The dimensions of the input distributions must match or the function will panic.
```
Comment (vladimir-ch) [inline stat/distmv/statdist.go]: This line belongs better to individual methods.

File: stat/distmv/statdist.go [hunk 4]
```diff
@@ -207,6 +207,76 @@ func (KullbackLeibler) DistUniform(l, r *Uniform) float64 {
...
+// The dimensions of the input distributions must match or DistNormal will panic.
+//
+// For two normal distributions, the Rényi divergence is computed as
+//  Σ_α = (1-α) Σ_l + αΣ_r
+//  D_α(l||r) = α/2 * (μ_l - μ_r)'*Σ_α^-1*(μ_l - μ_r) + 1/(2(α-1))*[ln(|Σ_λ|/(|Σ_l|^(1-α)+|Σ_r|^α)]
+//
+// For a more nicely formatted version of the formula, see Eq. 15 of
+//  Kolchinsky, Artemy, and Brendan D. Tracey. "Estimating Mixture Entropy
+//  with Pairwise Distances." arXiv preprint arXiv:1706.02419 (2017).
+// Note that the this formula is for the Chernoff divergence, which differs from
```
Comment (vladimir-ch) [inline stat/distmv/statdist.go]: s/the//

File: stat/distmv/statdist.go [hunk 5]
```diff
@@ -207,6 +207,76 @@ func (KullbackLeibler) DistUniform(l, r *Uniform) float64 {
...
+type Renyi struct {
+	Alpha float64
+}
+
+// DistNormal returns the Rényi divergence between normal distributions l and r.
+// The dimensions of the input distributions must match or DistNormal will panic.
+//
+// For two normal distributions, the Rényi divergence is computed as
+//  Σ_α = (1-α) Σ_l + αΣ_r
+//  D_α(l||r) = α/2 * (μ_l - μ_r)'*Σ_α^-1*(μ_l - μ_r) + 1/(2(α-1))*[ln(|Σ_λ|/(|Σ_l|^(1-α)+|Σ_r|^α)]
```
Comment (vladimir-ch) [inline stat/distmv/statdist.go]: There is a parenthesis missing for `ln` in the last term, Σ_λ should be Σ_α, and + should be *. Btw, with [15] you cite my former colleague from my faculty :-)

File: stat/distmv/statdist.go (hunk 3 above)
Comment (btracey) [inline stat/distmv/statdist.go]: Done.

File: stat/distmv/statdist.go (hunk 1 above)
Comment (btracey) [inline stat/distmv/statdist.go]: It's the integral over the whole domain, -inf to inf.

Done.

File: stat/distmv/statdist.go (hunk 5 above)
Comment (btracey) [inline stat/distmv/statdist.go]: This formula is pretty much impossible to write correctly I've decided.

File: stat/distmv/statdist.go (hunk 4 above)
Comment (btracey) [inline stat/distmv/statdist.go]: Done.

File: stat/distmv/statdist_test.go [hunk 6]
```diff
@@ -259,3 +259,72 @@ func klSample(dim, samples int, l RandLogProber, r LogProber) float64 {
...
+		want = KullbackLeibler{}.DistNormal(b, a)
+		got = Renyi{Alpha: 0.9999999}.DistNormal(b, a) // very close to 1 but not equal to 1.
+		if math.Abs(want-got) > 1e-6 {
+			t.Errorf("Case %d: Renyi mismatch with KL(b||a): got %v, want %v", cas, got, want)
+		}
+	}
+}
+
+// renyiSample finds an estimate of the Rényi divergence through sampling.
+// Note that this sampling procedure only works if l has broader support than
```
Comment (kortschak) [inline stat/distmv/statdist_test.go]: This r is lonely.

**Comment by btracey** [comment]
Hmm, I guess when we squash commits we need a new approval? The small change diffs I find useful when reviewing, so I don't want us to overly squash to get around this.

**Comment by btracey** [comment]
Re-approve Ping?

**Comment by kortschak** [comment]
If you use the squash option here, you don't lose the approval. This will only work if you are happy with squashing to a single commit - which is true here.

**Pull request merged**